test/unit/controller/StateController.spec.ts

```diff
@@ -0,0 +1,109 @@
+import { AccountCreationState } from '../../../src/state/AccountCreationState';
+import { AttachState } from '../../../src/state/AttachState';
+import { CleanUpState } from '../../../src/state/CleanUpState';
+import { CLIService } from '../../../src/services/CLIService';
+
+describe('StateController', () => {
+
+  describe('should transition to next state on Finish event', () => {
+    let stateController: StateController;
+    let serviceLocatorStub: sinon.SinonStub;
+    let cleanUpStateStub: sinon.SinonStubbedInstance<CleanUpState>;
+    let getStartConfigurationStub: sinon.SinonStub;
+
+    before(() => {
+      // Create a stub for the ServiceLocator
+      const stubbedInitState = sinon.createStubInstance(InitState);
+      const stubbedStartState = sinon.createStubInstance(StartState);
+      const stubbedNetworkPrepState = sinon.createStubInstance(NetworkPrepState);
+      const stubbedAccountCreationState = sinon.createStubInstance(AccountCreationState);
+      cleanUpStateStub = sinon.createStubInstance(CleanUpState);
+      const stubbedAttachState = sinon.createStubInstance(AttachState);
+      const loggerServiceStub = sinon.createStubInstance(LoggerService);
+  
+      serviceLocatorStub = sinon.stub(ServiceLocator.Current, 'get').returns(loggerServiceStub);
+  
+      getStartConfigurationStub = sinon.stub(StateData.prototype, <any>"getStartConfiguration").returns({
+          'stateMachineName' : 'start',
+          'states' : [
+            stubbedInitState,
+            stubbedStartState,
+            stubbedNetworkPrepState,
+            stubbedAccountCreationState,
+            cleanUpStateStub,
+            stubbedAttachState
+          ]
+      });
+      stateController = new StateController('start');
+    });
+
+    after(() => {
+      // Restore the original ServiceLocator after each test
+      cleanUpStateStub.onStart.restore();
+      serviceLocatorStub.restore();
+      getStartConfigurationStub.restore();
+    });
+
+    it('should transition to next state on Finish event', async () => {
+      // Starting the state machine
+      await stateController.startStateMachine();
+
+      // Sending two finish events
+      await stateController.update(EventType.Finish);
+      await stateController.update(EventType.Finish);
+
+      expect(stateController['currStateNum']).to.equal(2);
+    });
+
+  });
+
+  describe('should handle other events correctly', () => {
+    let stateController: StateController;
+    let getStartConfigurationStub: sinon.SinonStub;
+
+    before(() => {
+      const loggerServiceStub = sinon.createStubInstance(LoggerService);
+      const cliServiceStub = sinon.createStubInstance(CLIService);
+      
+      // Stub the first call to ServiceLocator.Current.get
+      const getStub1 = sinon.stub(ServiceLocator.Current, 'get').returns(loggerServiceStub);
+      //getStub1.onFirstCall().returns(loggerServiceStub);
+      getStub1.onSecondCall().returns(cliServiceStub);
```

Comment:
I genuinely dont get why this works but
`const getStub1 = sinon.stub(ServiceLocator.Current, 'get');
      getStub1.onFirstCall().returns(loggerServiceStub);
      getStub1.onSecondCall().returns(cliServiceStub);`
this does not it give an error that in the CleanUpState this.logger.trace is calling trace on undefined...
@georgi-l95 @stefan-stefanooov @yiliev0 
any ideas?
sinon is really confusing me